### PR TITLE
Add support for text stroke

### PR DESCRIPTION
**Pull request by Flashfyre**
Adds support for text stroke. Please note that I have not tested this directly. I have however made the same changes to a compiled version to know that it works. Regardless, it should be tested before being merged, but it's such a simple change that I doubt it will fail.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -30,6 +30,9 @@ export default class extends three.Sprite {
     this._fontFace = 'Arial';
     this._fontSize = 90; // defines text resolution
     this._fontWeight = 'normal';
+    
+    this._strokeWidth = 0;
+    this._strokeColor = 'transparent';
```
Comment (vasturiano) [inline src/index.js:35]: Better initialize this to `null` instead, to prevent doing string comparisons.

File: src/index.js [hunk 2]
```diff
@@ -116,13 +123,26 @@ export default class extends three.Sprite {
...
+    ctx.strokeStyle = this.strokeColor;
     ctx.textBaseline = 'bottom';
-
-    lines.forEach((line, index) => ctx.fillText(
-      line,
-      (innerWidth - ctx.measureText(line).width) / 2,
-      (index + 1) * this.fontSize
-    ));
+    
+    const drawTextStroke = this.strokeWidth && this.strokeColor !== 'transparent';
```
Comment (vasturiano) [inline src/index.js:130]: Remove string comparison on `strokeColor` if initialized as `null`.

File: src/index.js [hunk 3]
```diff
@@ -116,13 +123,26 @@ export default class extends three.Sprite {
...
+    
+    lines.forEach((line, index) => {
+      if (drawTextStroke) {
+        ctx.strokeText(
+          line,
+          (innerWidth - ctx.measureText(line).width) / 2,
```
Comment (vasturiano) [inline src/index.js:136]: Let's store `x` and `y` in block vars, to avoid code duplication.

File: src/index.js [hunk 4]
```diff
@@ -116,13 +123,26 @@ export default class extends three.Sprite {
...
-      line,
-      (innerWidth - ctx.measureText(line).width) / 2,
-      (index + 1) * this.fontSize
-    ));
+    
+    const drawTextStroke = this.strokeWidth && this.strokeColor !== 'transparent';
+    
+    lines.forEach((line, index) => {
+      if (drawTextStroke) {
+        ctx.strokeText(
```
Comment (vasturiano) [inline src/index.js:134]: Shouldn't we stroke **after** we fill?

**Comment by vasturiano** [comment]
That's great, thanks for the PR @Flashfyre!

Could you add the new attributes to the types definitions too:
https://github.com/vasturiano/three-spritetext/blob/master/src/index.d.ts

File: src/index.js (hunk 4 above)
Comment (Flashfyre) [reply src/index.js:134]: I don't believe so, that just overlaps the stroke on top of the fill meaning you can't see the intended fill colour.

Comment (Flashfyre) [reply src/index.js:136]: I suppose it depends on how thick your stroke is, but I can't think of a scenario where it would look better over top.

**Comment by vasturiano** [comment]
Thanks for the changes @Flashfyre!

I did some small fixes and also added docs and a quick example.
https://vasturiano.github.io/three-spritetext/example/basic/

**Comment by Flashfyre** [comment]
After implementing this update in my project, something is wrong, it is rendering a black background on my text...

**Comment by Flashfyre** [comment]
It appears to be unrelated to this change, and happened in 1.4.0 -> 1.5.0

**Comment by vasturiano** [comment]
Can you share a small code example to reproduce it?

**Comment by Flashfyre** [comment]
https://github.com/Flashfyre/Yume-2kki-Explorer/blob/3be1f90406053f82196f2fb766f0310184dc01c3/index.html#L1055-L1060

This here was fine in my modified version of 1.4.0 that added only the changes I made in this PR, so I essentially skipped any changes you made after that. In 1.5.0 it adds a black text background.

**Comment by Flashfyre** [comment]
I do use SpriteText in another location in this same file that may possibly be interfering with it in a way that only happens in 1.5.0+.

**Comment by Flashfyre** [comment]
I have to update to mention I was a bit confused in that I saw the merge and that it was 1.5.1, so I figured there was a separate change before my PR was merged, though this appears to not be the case so this change was the cause, though it was fine in my custom version [here](https://github.com/Flashfyre/Yume-2kki-Explorer/blob/master/public/js/three-spritetext-modded.js).

**Comment by Flashfyre** [comment]
Sorry for the confusion, but I have figured out the problem. For some reason, possibly due to some additional logic you added, the scaling for stroke width was completely different, and the width of 10 I used for my project that looked fairly thin was so thick in this one that it appeared to be a black background behind the text.

**Comment by vasturiano** [comment]
Yes, just reached the same conclusion. The `strokeWidth` is in units proportional to the text size, as mentioned in the docs. A value of `10` totally fills the canvas which is why you're seeing a black rectangle.

**Pull request merged**